### PR TITLE
add locale.rake

### PR DESCRIPTION
**Pull request by JasonYCHuang**
Solve travis CI problem in https://github.com/shakacode/react-webpack-rails-tutorial/pull/340

---
This change is [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/717)

### CODE REVIEW

**Comment by justin808** [comment]
@JasonYCHuang This looks fine. If we have this, then it's optional to add the locale building to the application startup. It would be really slick if we checked for the locale files being current at the same place migrations are confirmed on controller requests....

In order to merge this:

1. Doc updates (i18n file? README?)
2. Please see https://github.com/ruby/rake/issues/7 for the long description on the rake task. I'd put in a link to the i18n doc file in the GitHub repo.
3. Changelog entry

**Comment by JasonYCHuang** [comment]
@justin808 I will update documents.

I am confused about Travis CI failure. It only fails in Ruby2.3.1.
So far, I didn't find any clue.
Do you have suggestions on this?

**Comment by justin808** [comment]
Travis.ci is consistently failing for at least one ruby version and I just restart it for just that version.

**Comment by justin808** [comment]
@JasonYCHuang See my comment above: https://github.com/shakacode/react_on_rails/pull/717#issuecomment-280276326

Do you need any help from me? I'd like to wrap this one up this weekend!

**Comment by JasonYCHuang** [comment]
@justin808 Yes, I need some advice.

I am going to update [i18n doc](https://github.com/shakacode/react_on_rails/blob/master/docs/basics/i18n.md).
If we will use __rake task__, instead of `ReactOnRails::LocalesToJs.new` in `config/application.rb`, to trigger generating locale js files, where should I put this rake script.

In `package.json`, we can use `prebuild` hook to make sure this happens before all the build, but we have several build scripts, like `build:client`, `build:server`. It is not smart to set __rake task__ in all of them. Where should I put the task?

**Comment by JasonYCHuang** [comment]
@justin808 What do you mean: `Please see ruby/rake#7 for the long description on the rake task. I'd put in a link to the i18n doc file in the GitHub repo.`?

Should I add more information to `desc`? or only add a link to `i18n.md` pointing this rake file?

**Comment by justin808** [comment]
@JasonYCHuang I sent you invites to Screen Hero and Slack. Let's get this locale stuff resolved! Awesome work!

> Should I add more information to desc? or only add a link to i18n.md pointing this rake file?
The more docs the better! Remember this is some plain text. I type `rake -D` and get a good description of this!

Overall:

1. Running the build task in config/application.rb is a bit awkward for a production deploy, because the assets should have been precompiled.
2. Yes, I think the npm scripts should be responsible for ensuring that the rake task runs at the appropriate time. Ideally, during development, we'd have a "watch" version of that task. The alternative is to have a middleware check on requests to see if the locales are outdated. Even better would be if we could have webpack ensure this! All that being said, I think the locales files will change only occasionally outside of the main language. Most of that work is done by separate teams that will do this in a batch.


http://www.marcusoft.net/2015/08/pre-and-post-hooks-for-npm-scripting.html#pre-and-post-hooks-for-custom-script

**Comment by justin808** [comment]
This is where we fix up precompile by default:

https://github.com/shakacode/react_on_rails/blob/master/lib/tasks/assets.rake#L38

And this doc needs to be updated... as well as the new locale one.
https://github.com/shakacode/react_on_rails/blob/master/docs/additional-reading/heroku-deployment.md#more-details-on-precompilation-using-webpack-to-create-javascript-assets

File: lib/tasks/assets.rake [hunk 1]
```diff
@@ -50,10 +51,21 @@ Compile assets with webpack
...
       if ReactOnRails.configuration.npm_build_production_command.present?
         sh "cd client && #{ReactOnRails.configuration.npm_build_production_command}"
       end
     end
+
+    desc <<-DESC
+When "ReactOnRails.configuration.i18n_dir" is set, it indeicates that javascript locale files are needed.
+This task generates javascript locale files: `translations.js` & `default.js`.
+    DESC
+    desc "Generate i18n javascript files"
```
Comment (justin808) [inline lib/tasks/assets.rake]: Formatting is not right See example above in this file.

```ruby
 desc << Generate i18n javascript files
This task generates javascript locale files: `translations.js` & `default.js` and places them in
the "ReactOnRails.configuration.i18n_dir".
DESC
```

**Comment by justin808** [comment]
Tiny comments. @robwise Please give me the OK on this one.

---

Reviewed 2 of 2 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[docs/basics/i18n.md, line 58 at r3](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-KdbNRKJD8pnnymGH9kN:-KdbNRKJD8pnnymGH9kO:b1mkw7m) ([raw file](https://github.com/shakacode/react_on_rails/blob/822509aecba0386b9ccf3408f97b0ad53c965e11/docs/basics/i18n.md#L58)):*
> ```Markdown
>     "scripts": {
>       ...
>       // Assuem that the following three lines are your original build script.
> ```

assume
two lines

---

*[docs/basics/i18n.md, line 69 at r3](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-KdbN_Nnvjv1o-x09fAd:-KdbN_Nnvjv1o-x09fAe:b-s84ihj) ([raw file](https://github.com/shakacode/react_on_rails/blob/822509aecba0386b9ccf3408f97b0ad53c965e11/docs/basics/i18n.md#L69)):*
> ```Markdown
>   ```
> 
>   For __dev__, add `bundle exec rake react_on_rails:locale` to `Procfile.dev` before `npm build`. 
> ```

This is confusing if there's server side rendering...as there's 2 npm builds there

and is this command in-line?

ahh you show below

say before `npm build` for the client bundle

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-:-KdbOG9Mzj8tR7He7rHk:bg9uhk5)*

**Comment by JasonYCHuang** [comment]
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[docs/basics/i18n.md, line 58 at r3](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-KdbNRKJD8pnnymGH9kN:-KdbadKkHU8Z1xBcN6xw:b-7rl0kd) ([raw file](https://github.com/shakacode/react_on_rails/blob/822509aecba0386b9ccf3408f97b0ad53c965e11/docs/basics/i18n.md#L58)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

assume
two lines
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/717)*

**Comment by JasonYCHuang** [comment]
done

---

Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[docs/basics/i18n.md, line 69 at r3](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-KdbN_Nnvjv1o-x09fAd:-KdbfYJF4J2qraeNklfF:byzhs2n) ([raw file](https://github.com/shakacode/react_on_rails/blob/822509aecba0386b9ccf3408f97b0ad53c965e11/docs/basics/i18n.md#L69)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

This is confusing if there's server side rendering...as there's 2 npm builds there

and is this command in-line?

ahh you show below

say before `npm build` for the client bundle
</blockquote></details>

yes.

The rake task should be placed before the 1st npm build.

---

*[lib/tasks/assets.rake, line 64 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-KdQmxy4sYLPRPxy5cma-r1-64:-KdbfWj-62kJYdtvfvjn:bb7f2mf) ([raw file](https://github.com/shakacode/react_on_rails/blob/724eba872b9999d1411f804fe243fad7687651f9/lib/tasks/assets.rake#L64)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

Formatting is not right See example above in this file.

```ruby
 desc << Generate i18n javascript files
This task generates javascript locale files: `translations.js` & `default.js` and places them in
the "ReactOnRails.configuration.i18n_dir".
DESC
``` 
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-:-KdbfhaRHyyF-2bZHJDs:bb7f2mf)*

**Comment by justin808** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[docs/basics/i18n.md, line 58 at r3](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-KdbNRKJD8pnnymGH9kN:-KdblFh-CfqCWXeTLnix:b-62jjts) ([raw file](https://github.com/shakacode/react_on_rails/blob/822509aecba0386b9ccf3408f97b0ad53c965e11/docs/basics/i18n.md#L58)):*
<details><summary><i>Previously, JasonYCHuang (Jason Huang) wrote…</i></summary><blockquote>

done
</blockquote></details>

assume is still mispelled

and in the react-webpack-rails-tutorial

you do a `prebuild:dev`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/717)*

**Comment by justin808** [comment]
close

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-:-KdblQEhTf4qw5XMXfuM:b-7s6oz1)*

**Comment by JasonYCHuang** [comment]
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[docs/basics/i18n.md, line 58 at r3](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-KdbNRKJD8pnnymGH9kN:-KdeAPhjb-23YzoOeNaT:bvnrpcm) ([raw file](https://github.com/shakacode/react_on_rails/blob/822509aecba0386b9ccf3408f97b0ad53c965e11/docs/basics/i18n.md#L58)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

assume is still mispelled

and in the react-webpack-rails-tutorial

you do a `prebuild:dev`
</blockquote></details>

updated

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/717)*

**Comment by JasonYCHuang** [comment]
No pre-hook needed in `package.json`

1. to get js locale files in dev: https://github.com/shakacode/react_on_rails/pull/717/files#diff-51a05a5d3f0a0c62b9452b7596e2a82cR2

2. to get js locale files in test & production: https://github.com/shakacode/react_on_rails/pull/717/files#diff-52cfe83d7f9f1c952a7995dfb0f99341R53

3. Since `config.i18n_dir` will be the key to decide whether to generate locale files or not, I comment it for default. When users need i18n, they need to uncomment it. https://github.com/shakacode/react_on_rails/pull/717/files#diff-391d7a2ed038bee9f49a7e7adefc133eR66

---

Review status: 1 of 6 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-:-Kdzkq_qzKfP0wuNOQX_:bsy64wc)*

**Comment by justin808** [comment]
Very nice. 

For testing, we should have a call to generate the i18n task here:

https://github.com/shakacode/react_on_rails/blob/master/lib/react_on_rails/test_helper/ensure_assets_compiled.rb#L29

---

Reviewed 1 of 1 files at r5, 3 of 5 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[CHANGELOG.md, line 10 at r7](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-Ke2xLJ9oLi4GFx_jHe7:-Ke2xLJ9oLi4GFx_jHe8:b-dbfdbx) ([raw file](https://github.com/shakacode/react_on_rails/blob/cb75a4f5975280e85090dfde74117d6d3c6027e5/CHANGELOG.md#L10)):*
> ```Markdown
> 
> ### Added
> - Allow using rake task to generate javascript locale files. [#717](https://github.com/shakacode/react_on_rails/pull/717) by [JasonYCHuang](https://github.com/JasonYCHuang).
> ```

Allow using rake task to generate javascript locale files. The test helper automatically creates the localization files when needed.

---

*[docs/basics/i18n.md, line 49 at r7](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-Ke2vuG0y6IiDjmY_iNX:-Ke2vuG0y6IiDjmY_iNY:b-arqzeg) ([raw file](https://github.com/shakacode/react_on_rails/blob/cb75a4f5975280e85090dfde74117d6d3c6027e5/docs/basics/i18n.md#L49)):*
> ```Markdown
> 4. Javascript locale files must be generated before `yarn build`. 
>   
>   Once you setup `config.i18n_dir` as in the previous step, react_on_rails will help you to do this.
> ```

for testing and for production deployments. For development, you should adjust your Procfiles so that they run bundle exec rake react_on_rails:locale` before running the `yarn run build:development` (webpack watch process)

---

*[lib/generators/react_on_rails/templates/base/base/Procfile.dev.tt, line 2 at r7](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-Ke2wYKwXOWXD4HlLhtz:-Ke2wYKx_ld1NzTeGADB:b-a8llct) ([raw file](https://github.com/shakacode/react_on_rails/blob/cb75a4f5975280e85090dfde74117d6d3c6027e5/lib/generators/react_on_rails/templates/base/base/Procfile.dev.tt#L2)):*
> ```
> web: rails s -p 3000
> client: sh -c 'rm app/assets/webpack/* || true && cd client && bundle exec rake react_on_rails:locale && yarn run build:development'
> ```

nice! Did you try the generator?

---

*[lib/tasks/assets.rake, line 53 at r7](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-Ke2wfV5d7XSNS3AeY1e:-Ke2wfV5d7XSNS3AeY1f:b-83sjoh) ([raw file](https://github.com/shakacode/react_on_rails/blob/cb75a4f5975280e85090dfde74117d6d3c6027e5/lib/tasks/assets.rake#L53)):*
> ```Ruby
> sh "cd client && `ReactOnRails.configuration.npm_build_production_command`"
>     DESC
>     task webpack: :locale do
> ```

Nice!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-:-Ke2whSCT59XQVHXTeLo:bx7p8o4)*

**Comment by JasonYCHuang** [comment]
done

---

Review status: 4 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[CHANGELOG.md, line 10 at r7](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-Ke2xLJ9oLi4GFx_jHe7:-Ke4SDJ_Ki0gouaS4JbR:bb7f2mf) ([raw file](https://github.com/shakacode/react_on_rails/blob/cb75a4f5975280e85090dfde74117d6d3c6027e5/CHANGELOG.md#L10)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

Allow using rake task to generate javascript locale files. The test helper automatically creates the localization files when needed.
</blockquote></details>

done

---

*[docs/basics/i18n.md, line 49 at r7](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-Ke2vuG0y6IiDjmY_iNX:-Ke4SEdEO6sWbTJSTTYG:bb7f2mf) ([raw file](https://github.com/shakacode/react_on_rails/blob/cb75a4f5975280e85090dfde74117d6d3c6027e5/docs/basics/i18n.md#L49)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

for testing and for production deployments. For development, you should adjust your Procfiles so that they run bundle exec rake react_on_rails:locale` before running the `yarn run build:development` (webpack watch process)
</blockquote></details>

done

---

*[lib/generators/react_on_rails/templates/base/base/Procfile.dev.tt, line 2 at r7](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-Ke2wYKwXOWXD4HlLhtz:-Ke4SFmfBJ6WmijLYjMJ:biwvsul) ([raw file](https://github.com/shakacode/react_on_rails/blob/cb75a4f5975280e85090dfde74117d6d3c6027e5/lib/generators/react_on_rails/templates/base/base/Procfile.dev.tt#L2)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

nice! Did you try the generator?
</blockquote></details>

yes, I create a new rails project and install react_on_rails. It looks good.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-:-Ke4SAHVul0Y-X1acgOh:bb7f2mf)*

**Comment by justin808** [comment]
:lgtm:

@jasonychuang Please confirm that you manually verified checking the locales files works in the test helper and I'll merge.

---

Reviewed 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[lib/generators/react_on_rails/templates/base/base/Procfile.dev.tt, line 2 at r7](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-Ke2wYKwXOWXD4HlLhtz:-Ke59Eo-rYCiSbjpvSsL:bkbjs3e) ([raw file](https://github.com/shakacode/react_on_rails/blob/cb75a4f5975280e85090dfde74117d6d3c6027e5/lib/generators/react_on_rails/templates/base/base/Procfile.dev.tt#L2)):*
<details><summary><i>Previously, JasonYCHuang (Jason Huang) wrote…</i></summary><blockquote>

yes, I create a new rails project and install react_on_rails. It looks good.
</blockquote></details>

fabulous.

---

*[lib/react_on_rails/test_helper/ensure_assets_compiled.rb, line 30 at r8](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-Ke595cYaYO68sk-PDur:-Ke595cYaYO68sk-PDus:b2e3i0j) ([raw file](https://github.com/shakacode/react_on_rails/blob/a10e4db97483ff5dc57fd41f8bca92dc1ba762cd/lib/react_on_rails/test_helper/ensure_assets_compiled.rb#L30)):*
> ```Ruby
>         self.class.has_been_run = true
> 
>         ReactOnRails::LocalesToJs.new if ReactOnRails.configuration.i18n_dir.present?
> ```

This is awesome.

@robwise are you OK with us not adding a test here: https://github.com/shakacode/react_on_rails/blob/master/spec/react_on_rails/test_helper/ensure_assets_compiled_spec.rb#L8

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-:-Ke59CN3YS3eeFivPKR5:b-b2haqq)*

**Comment by robwise** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[lib/react_on_rails/test_helper/ensure_assets_compiled.rb, line 30 at r8](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-Ke595cYaYO68sk-PDur:-Ke5NHtpDXT7PDfqK2tC:balodb0) ([raw file](https://github.com/shakacode/react_on_rails/blob/a10e4db97483ff5dc57fd41f8bca92dc1ba762cd/lib/react_on_rails/test_helper/ensure_assets_compiled.rb#L30)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

This is awesome.

@robwise are you OK with us not adding a test here: https://github.com/shakacode/react_on_rails/blob/master/spec/react_on_rails/test_helper/ensure_assets_compiled_spec.rb#L8
</blockquote></details>

I would definitely write a test if you are introducing new logic

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/717)*

**Comment by JasonYCHuang** [comment]
Yes, I manually run `ReactOnRails::TestHelper::EnsureAssetsCompiled.new.call` under rails console. Local files will be generated when `config.i18n_dir` is set.

---

Review status: 7 of 8 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/717#-:-Ke5gj-cdriiLP398PT-:b5nifre)*

**Pull request merged**